File: python/examples/verify_route.py
"""Verify both hops and interpret the result.

    ANONROUTER_API_KEY=ar_... python examples/verify_route.py

Optional env:
    ANONROUTER_BASE_URL   default https://api.anonrouter.ai
    ANONROUTER_MODEL      default venice-uncensored
    ANONROUTER_PROVIDER   default venice

Prints what was established, what was not, and why, then exits non-zero if the
route did not meet the threshold. Attestation only: not a billable inference call.
"""

from __future__ import annotations

import os
import sys

from anonrouter_confidential import at_least, create_client, describe_state

BASE_URL = os.environ.get("ANONROUTER_BASE_URL", "https://api.anonrouter.ai")
MODEL = os.environ.get("ANONROUTER_MODEL", "venice-uncensored")
PROVIDER = os.environ.get("ANONROUTER_PROVIDER", "venice")

# The bar this program insists on. `policy_matched` would accept a verdict resting
# on a document we did not produce, which is weaker than most callers mean.
REQUIRED = "cryptographically_checked"


def main() -> int:
    api_key = os.environ.get("ANONROUTER_API_KEY")
    if not api_key:
        print("Set ANONROUTER_API_KEY to your AnonRouter API key (inference scope).", file=sys.stderr)
        return 2

    with create_client(BASE_URL, api_key=api_key) as client:
        verdict = client.verify_route(
            model=MODEL,
            provider=PROVIDER,
            # Ask about AnonRouter's own plane too. Against a deployment not
            # running in a CVM this reports `unavailable`: untrusted, not a pass.
            gateway={"allow_candidate_policy": True},
        )

    print(f"route      {verdict.route.provider}/{verdict.route.model}")
    visible = (
        "  (AnonRouter sees your plaintext on this route)"
        if verdict.content_visible_to_anonrouter
        else ""
    )
    print(f"modality   {verdict.route.privacy_modality}{visible}")
    print()

    for name, hop in (("hop 1 gateway ", verdict.gateway), ("hop 2 provider", verdict.provider)):
        print(f"{name}  {hop.state if hop.requested else 'not requested'}")
        if hop.requested:
            print(f"                {describe_state(hop.state)}")
            if hop.failed_checks:
                print(f"                failed: {', '.join(hop.failed_checks)}")
            if hop.advisory_gaps:
                print(f"                gaps:   {', '.join(hop.advisory_gaps)}")

    print()
    print(f"overall    {verdict.overall_state}")
    for m in verdict.binding_mismatches:
        # Two honestly-attested parties on the wrong route is still the wrong route.
        print(f"MISMATCH   {m.field_name}: asked for {m.expected}, {m.source} reported {m.observed}")
    if verdict.reason:
        print(f"reason     {verdict.reason}")

    if not at_least(verdict.overall_state, REQUIRED):
        print(f"\nFAILED: route did not reach {REQUIRED}.", file=sys.stderr)
        print("See VERIFYING.md for what each failed check means.", file=sys.stderr)
        return 1
    print(f"\nOK: route reached {REQUIRED}.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: js/confidential/examples/verify-route.ts
// Verify both hops and interpret the result.
//
//   npx tsx examples/verify-route.ts
//
// Reads ANONROUTER_BASE_URL, ANONROUTER_API_KEY, and optionally
// ANONROUTER_MODEL / ANONROUTER_PROVIDER. Prints what was established, what was
// not, and why, then exits non-zero if the route did not meet the threshold.

import "./_env.js";
import { createClient, atLeast, describeState, type RouteVerificationState } from "../src/index.js";
import { createAnonRouterDcapVerifier } from "../src/gateway/dcap/index.js";

const BASE_URL = process.env.ANONROUTER_BASE_URL ?? "https://api.private.anonrouter.ai";
const CONTROL_URL = process.env.ANONROUTER_CONTROL_URL ?? "https://api.anonrouter.ai";
const MODEL = process.env.ANONROUTER_MODEL ?? "venice-uncensored";
const PROVIDER = process.env.ANONROUTER_PROVIDER ?? "venice";

// The bar this program insists on. `policy_matched` would accept a verdict that
// rests on a document we did not produce, which is weaker than most callers mean.
const REQUIRED: RouteVerificationState = "cryptographically_checked";

const apiKey = process.env.ANONROUTER_API_KEY;
if (!apiKey) {
  console.error("Set ANONROUTER_API_KEY to your AnonRouter API key (inference scope).");
  process.exit(2);
}

const client = createClient({ baseUrl: BASE_URL, controlBaseUrl: CONTROL_URL, apiKey });

const verdict = await client.verifyRoute({
  model: MODEL,
  provider: PROVIDER,
  // Ask about AnonRouter's own plane too and chain its quote to Intel.
  gateway: { chainVerifier: createAnonRouterDcapVerifier() }
});

console.log(`route      ${verdict.route.provider}/${verdict.route.model}`);
console.log(`modality   ${verdict.route.privacyModality}`
  + (verdict.contentVisibleToAnonRouter ? "  (AnonRouter sees your plaintext on this route)" : ""));
console.log();

for (const [name, hop] of [["hop 1 gateway ", verdict.gateway], ["hop 2 provider", verdict.provider]] as const) {
  const scope = hop.requested ? hop.state : "not requested";
  console.log(`${name}  ${scope}`);
  if (hop.requested) {
    console.log(`                ${describeState(hop.state)}`);
    if (hop.failedChecks.length > 0) console.log(`                failed: ${hop.failedChecks.join(", ")}`);
    if (hop.advisoryGaps.length > 0) console.log(`                gaps:   ${hop.advisoryGaps.join(", ")}`);
  }
}

console.log();
console.log(`overall    ${verdict.overallState}`);
if (verdict.bindingMismatches.length > 0) {
  // Two honestly-attested parties on the wrong route is still the wrong route.
  for (const m of verdict.bindingMismatches) {
    console.log(`MISMATCH   ${m.field}: asked for ${m.expected}, ${m.source} reported ${m.observed}`);
  }
}
if (verdict.reason) console.log(`reason     ${verdict.reason}`);

if (!atLeast(verdict.overallState, REQUIRED)) {
  console.error(`\nFAILED: route did not reach ${REQUIRED}.`);
  console.error("See VERIFYING.md for what each failed check means.");
  process.exit(1);
}
console.log(`\nOK: route reached ${REQUIRED}.`);
